agent/reconnectingpty/buffered.go

```diff
@@ -0,0 +1,238 @@
+				// not found for example).
+				rpty.Close("unable to read pty output, command might have exited")
+				break
+			}
+			part := buffer[:read]
+			rpty.mutex.Lock()
+			_, err = rpty.circularBuffer.Write(part)
+			if err != nil {
+				logger.Error(ctx, "write to circular buffer", slog.Error(err))
+				rpty.metrics.WithLabelValues("write_buffer").Add(1)
+			}
+			// TODO: Instead of ranging over a map, could we send the output to a
+			// channel and have each individual Attach read from that?
+			for cid, conn := range rpty.activeConns {
+				_, err = conn.Write(part)
+				if err != nil {
+					logger.Warn(ctx,
+						"error writing to active connection",
+						slog.F("connection_id", cid),
+						slog.Error(err),
+					)
+					rpty.metrics.WithLabelValues("write").Add(1)
+				}
+			}
+			rpty.mutex.Unlock()
+		}
+	}()
+
+	return rpty
+}
+
+// lifecycle manages the lifecycle of the reconnecting pty.  If the context ends
+// the reconnecting pty will be closed.
+func (rpty *bufferedReconnectingPTY) lifecycle(ctx context.Context, logger slog.Logger) {
+	rpty.timer = time.AfterFunc(attachTimeout, func() {
+		rpty.Close("reconnecting pty timeout")
+	})
+
+	logger.Debug(ctx, "reconnecting pty ready")
+	rpty.state.setState(StateReady, nil)
+
+	state, reasonErr := rpty.state.waitForStateOrContext(ctx, StateClosing)
+	if state < StateClosing {
+		// If we have not closed yet then the context is what unblocked us (which
+		// means the agent is shutting down) so move into the closing phase.
+		rpty.Close(reasonErr.Error())
+	}
+	rpty.timer.Stop()
+
+	rpty.mutex.Lock()
+	defer rpty.mutex.Unlock()
+
+	rpty.circularBuffer.Reset()
+
+	// Log these closes only for debugging since the connections or processes
+	// might have already closed on their own.
+	for _, conn := range rpty.activeConns {
+		err := conn.Close()
+		if err != nil {
+			logger.Debug(ctx, "closed conn with error", slog.Error(err))
+		}
+	}
+
+	err := rpty.ptty.Close()
+	if err != nil {
+		logger.Debug(ctx, "closed ptty with error", slog.Error(err))
+	}
+
+	err = rpty.process.Kill()
+	if err != nil {
+		logger.Debug(ctx, "killed process with error", slog.Error(err))
+	}
+
+	logger.Debug(ctx, "closed reconnecting pty")
+	rpty.state.setState(StateDone, xerrors.Errorf("reconnecting pty closed: %w", reasonErr))
+}
+
+func (rpty *bufferedReconnectingPTY) Attach(ctx context.Context, connID string, conn net.Conn, height, width uint16, logger slog.Logger) error {
+	// This will kill the heartbeat once we hit EOF or an error.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	logger.Debug(ctx, "reconnecting pty attach")
+	err := rpty.doAttach(ctx, connID, conn, height, width, logger)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		rpty.mutex.Lock()
+		defer rpty.mutex.Unlock()
+		delete(rpty.activeConns, connID)
+	}()
+
+	// Pipe conn -> pty and block.  pty -> conn is handled in newBuffered().
+	readConnLoop(ctx, conn, rpty.ptty, rpty.metrics, logger)
+	return nil
```

Comment:
Your comment on `Attach()` says:

> If the context ends or the process dies the connection will be detached and closed.

But I don't think that it really true.

Firstly, if the `ctx` passed in ends after we are `Ready`, I don't think there is any direct mechanism to detect that.  `readConnLoop` only uses the context for logging; it doesn't directly check whether the context expires.

Now, you might think: well, if the context expires, then the `lifecycle()` will close the connection, which will trigger an EOF when `readConnLoop` reads from it.  But!, that's the context passed to `newBuffered()`, **not** the context passed to `Attach()`.  As it happens, the agent uses the same context for both, but if we're depending on that behavior, then `Attach()` shouldn't accept a context, and we should just save the context passed to `newBuffered()` on the struct.

Secondly, if we go through the deferred function on L184, then the connection will not be closed.  This can happen if the `ptty` dies and `readConnLoop` returns before the `lifecycle()` can lock the mutex.

I don't think it's strictly necessary for `Attach()` to close the connections, since `handleReconnectingPTY` closes the conn in a defer, but we shouldn't claim on the interface that we will close it.

Reply:
I tweaked things to close the connection in `Attach()` to fix the issue with adding to the map after closing, and I went ahead and did it on context end as well so if the caller cancels their context then we break out of the loop. 

But having to close the connection to make `readConnLoop` yield is I think just an implementation detail and it still makes sense to avoid claiming `Attach()` closes the connection, so I left that part out of the comment!

Maybe it would be better to handle the context end in `readConnLoop` explicitly instead of relying on closing the connection to terminate that loop though, and this would avoid the potentially useless debug log when a connection is already closed.

Reply:
Ah nevermind I am not seeing a way to make the connection read abort when the context ends, so I guess we have to close it to make it yield.

Reply:
If you want to Read, but also abort on context you can do your reading in a separate goroutine, and send the results (read data and errors) over channels.  In the main goroutine you `select` on  `<-context.Done()` or the results channels.

The tricky bit is that you also need to wrap the sending on channels in a `select` that includes `<-context.Done()` so that you don't leak the Reader goroutine if the context ends and the main goroutine exits.

That's a lot of complexity, so I'm fine with closing the channel...

Reply:
Ahhhhhhh I see what I missed, I was too fixated on making the read itself abort, but if we have the read in a goroutine then it would not block `readConnLoop`, and once we yield since the agent closes the connection it would then make the read yield and the read goroutine would exit.

I will go ahead and keep closing it for now, but thank you for the explanation!